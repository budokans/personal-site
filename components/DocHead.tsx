import Head from "next/head";
import { ReactElement } from "react";
import { SiteMetadata } from "types";

interface DocHeadProps {
  readonly metadata: SiteMetadata;
}

export const DocHead = ({ metadata }: DocHeadProps): ReactElement => (
  <Head>
    <title>{metadata.title}</title>
    <meta name="description" content={metadata.description} />

    <meta property="og:title" content={metadata.title} />
    <meta property="og:description" content={metadata.description} />
    <meta property="og:image" content={`${metadata.canonical}/og.png`} />
    <meta property="og:locale" content="en_US" />
    <meta property="og:type" content="website" />
    <meta property="og:site_name" content="Steven Webster" />

    <meta name="twitter:card" content="summary_large_image" />
    <meta name="twitter:title" content={metadata.title} />
    <meta name="twitter:image:src" content={`${metadata.canonical}/og.png`} />
    <meta name="twitter:description" content={metadata.description} />
    <meta name="twitter:url" content={metadata.canonical} />

    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
    <link rel="manifest" href="/site.webmanifest" />
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
  </Head>
);
