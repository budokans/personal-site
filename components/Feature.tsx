import { Box } from "@chakra-ui/layout";
import { MotionBox } from "./MotionBox";
import SimpleBar from 'simplebar-react';
import { CloseButton } from "@chakra-ui/close-button";
import { useFeatureContext } from "../lib/featureContext";
import { AnimatePresence } from "framer-motion";

type FeatureProps = {
  project: number,
}

const containerVariants = {
  hidden: {
    y: "100vh"
  },
  visible: {
    y: "5vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      when: "beforeChildren",
      staggerChildren: 0.4,
    },
  },
  exit: {
    y: "100vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      duration: 5
    }
  }
}

const Feature = () => {
  const {projectToFeature, showFeature, closeFeature} = useFeatureContext();

  return (
    <AnimatePresence>
      { showFeature && (
        <>
          <Box position="absolute" top={0} right={0} bottom={0} left={0} bg="gray.100" />
      
          <MotionBox position="fixed" left={0} right={0} bottom={0} maxWidth="930px" marginY="0" marginX="auto" height="100vh"  variants={containerVariants} initial="hidden" animate="visible" exit="exit">
            
            <Box height="full" bg="white" borderWidth="1px" borderColor="gray.300" borderStyle="solid" borderRadius="xl">
              
              <SimpleBar style={{ maxHeight: '100%' }}>
                <CloseButton size="md" fontSize="16px" borderRadius="full" color="gray.900" bg="gray.100" p={0} top="10px" right="10px" onClick={closeFeature}  />
              </SimpleBar>
            </Box>
          </MotionBox> 
        </>
      )}
    </AnimatePresence>
  )
}

export default Feature;
