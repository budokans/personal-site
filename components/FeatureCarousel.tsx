import { Img } from "@chakra-ui/image";
import { Box, Stack } from "@chakra-ui/layout"
import { MotionBox } from "./MotionBox";

type FeatureCarouselProps = {
  media: string[]
}

const hideScrollbar = {
  '&::-webkit-scrollbar': {
    "width": "0 !important",
    "display": "none", 
  },
}

const containerVariants = {
  hidden: {
    x: "100vw",
  },
  visible: {
    x: 0,
    transition: {
      type: "spring",
      damping: 20,
    }
  }
}

const FeatureCarousel = ({media}: FeatureCarouselProps) => {
  return (
    <Box overflow="hidden" px={[4, 9]}>
      <MotionBox variants={containerVariants} initial="hidden" animate="visible" >

        <Stack 
          direction="row"
          alignItems="center"
          spacing={4} 
          height="calc(45vw + 35px)" 
          maxH="490px" 
          overflowX="scroll" 
          whiteSpace="nowrap" 
          sx={{ ...hideScrollbar, scrollSnapType: "x mandatory" }} 
          cursor="grab"
        >

        { media.map((image, idx) => {
          return (
            <Box 
              key={idx} 
              bg="gray.400" 
              borderRadius="xl"
              height="100%"
              minWidth="calc(70vw + 35px)"
              sx={{ scrollSnapAlign: "start", scrollPadding: "1.75rem" }} 
              p="7.5vw"
            >

              <Box h="100%" w="100%" >        
                <Img src={image} objectFit="cover" h="100%" w="100%" />
              </Box>
              
            </Box>
          )
        })}

        </Stack>
      </MotionBox>
    </Box>
  )
}

export default FeatureCarousel;