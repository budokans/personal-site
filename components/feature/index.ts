export * as Feature from "./Feature";
export * as Header from "./Header";
export * as Links from "./Links";
export * as Tech from "./Tech";
export * as Description from "./Description";
export { CarouselMobile } from "./CarouselMobile";
export { CarouselDesktop } from "./CarouselDesktop";
