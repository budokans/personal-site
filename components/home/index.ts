export * as Header from "./Header";
export * from "./Home";
export * as Portfolio from "./Portfolio";
