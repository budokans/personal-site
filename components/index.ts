export { DocHead } from "./DocHead";
export * as Feature from "./feature";
export * as Home from "./home";
export * as Link from "./Link";
export * as Motion from "./Motion";
