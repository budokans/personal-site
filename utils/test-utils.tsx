import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import { ChildrenProps } from "types";
import { ReactElement } from "react";

const withProviders = ({ children }: ChildrenProps): ReactElement => {
  return <ChakraProvider>{children}</ChakraProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: withProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
