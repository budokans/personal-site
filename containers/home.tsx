import { ReactElement } from "react";
import { SiteMetadata, Project } from "../types";
import { Header, Home, Portfolio } from "../components/home";
import { TextLink } from "../components/Link";

interface HomeContainerProps {
  readonly projects: readonly Project[];
  readonly metadata: SiteMetadata;
  readonly onPortfolioClick: (id: string) => void;
}

export const HomeContainer = ({
  projects,
  metadata,
  onPortfolioClick,
}: HomeContainerProps): ReactElement => {
  return (
    <Home.Home>
      <Header.Header>
        <Header.Heading>
          Hi, I&apos;m Steven. I&apos;m a full-stack developer who&apos;s helped
          build <TextLink url="https://rico.nz">Rico</TextLink> and{" "}
          <TextLink url="https://www.pivot-pointe.com">Pivot Pointe</TextLink>.
          I&apos;d love to help you with your software goals. Get in touch!
        </Header.Heading>

        <Header.HeaderContact>
          <Header.HeaderLocation>üìç {metadata.location}</Header.HeaderLocation>
          <Header.HeaderLinks contacts={metadata.contacts} />
        </Header.HeaderContact>
      </Header.Header>

      <Portfolio.Portfolio>
        {projects.map((project, idx) => {
          return (
            <Portfolio.PortfolioItem
              onPortfolioClick={onPortfolioClick}
              id={project.id}
              key={idx}
            >
              <Portfolio.PortfolioImage
                src={project.icon}
                alt={project.title}
              />
              <Portfolio.PortfolioInner
                idx={idx}
                projectsCount={projects.length}
              >
                <Portfolio.PortfolioTitle>
                  {project.title}
                </Portfolio.PortfolioTitle>
                <Portfolio.PortfolioText>
                  {project.shortBlurb}
                </Portfolio.PortfolioText>
              </Portfolio.PortfolioInner>
            </Portfolio.PortfolioItem>
          );
        })}
      </Portfolio.Portfolio>
    </Home.Home>
  );
};
