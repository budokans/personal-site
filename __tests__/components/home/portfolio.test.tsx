import { render, screen } from "test-utils";
import userEvent from "@testing-library/user-event";
import matchMediaPolyfill from "mq-polyfill";
import { Portfolio } from "components/home";

describe("<Portfolio />", () => {
  beforeAll(() => {
    matchMediaPolyfill(window);
    window.resizeTo = function resizeTo(width, height) {
      Object.assign(this, {
        innerWidth: width,
        innerHeight: height,
        outerWidth: width,
        outerHeight: height,
      }).dispatchEvent(new this.Event("resize"));
    };
  });

  test("<Portfolio.Item /> renders a button and calls openFeature() on click", () => {
    const openFeature = jest.fn();
    const { container } = render(
      <Portfolio.Item onPortfolioClick={openFeature} id="test">
        Children
      </Portfolio.Item>
    );

    expect(screen.getByRole("button")).toBeInTheDocument();
    userEvent.click(screen.getByRole("button"));
    expect(openFeature).toHaveBeenCalledTimes(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Porfolio.Image /> renders an image with populated data from props", () => {
    const project = {
      title: "Gretsch Geeks",
      icon: "/images/gretsch-logo.jpg",
    };

    const { container } = render(
      <Portfolio.Image src={project.icon} alt={project.title} />
    );

    expect(screen.getByAltText(/Gretsch Geeks icon/i)).toBeInTheDocument();
    expect(
      (container.querySelector("img") as HTMLElement).getAttribute("src")
    ).toBe("/images/gretsch-logo.jpg");
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Title /> renders with populated data from props", () => {
    const project = {
      title: "Gretsch Geeks",
    };

    const { container } = render(
      <Portfolio.Title>{project.title}</Portfolio.Title>
    );

    expect(screen.getByText(/Gretsch Geeks/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Text /> renders with populated data from props", () => {
    const project = {
      shortBlurb:
        "An example E-commerce buy-sell platform built with modern React, Apollo 3 and Keystone.js as a headless CMS.",
    };

    const { container } = render(
      <Portfolio.Text>{project.shortBlurb}</Portfolio.Text>
    );

    expect(screen.getByText(/Apollo 3/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Button /> renders a button with text", () => {
    const { container } = render(<Portfolio.Button />);

    expect(screen.getByRole(/button/i)).toBeInTheDocument();
    expect(screen.getByText(/View/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("<Portfolio.Inner />", () => {
  test("<Portfolio.Inner /> renders with a border-bottom if viewport width < 768px and projectsCount is 1", () => {
    window.resizeTo(767, 1000);

    const numberOfProjects = 1;
    const indexInArray = 0;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Inner /> renders without a border-bottom if viewport width < 768px, projectsCount is 2 and it is the last project", () => {
    window.resizeTo(767, 1000);

    const numberOfProjects = 2;
    const indexInArray = 1;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).not.toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Inner /> renders with a border-bottom if viewport width < 768px, projectsCount is 2 and it is not the last project", () => {
    window.resizeTo(767, 1000);

    const numberOfProjects = 2;
    const indexInArray = 0;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Inner /> renders with a border-bottom if viewport width > 768px and projectsCount is < 3", () => {
    window.resizeTo(768, 1000);

    const numberOfProjects = 2;
    const indexInArray = 0;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Inner /> renders without a border-bottom if viewport width > 768px, projectsCount is 3 and it is on the bottom row", () => {
    window.resizeTo(768, 1000);

    const numberOfProjects = 3;
    const indexInArray = 2;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).not.toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Inner /> renders with a border-bottom if viewport width > 768px, it is not in the first row, and it is the last one to be rendered before wrapping to a new row", () => {
    window.resizeTo(768, 1000);

    const numberOfProjects = 5;
    const indexInArray = 3;

    const { container } = render(
      <Portfolio.Inner
        idx={indexInArray}
        projectsCount={numberOfProjects}
      ></Portfolio.Inner>
    );

    expect(container.firstChild).toHaveStyle(
      `border-bottom: 1px solid lightgrey`
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
