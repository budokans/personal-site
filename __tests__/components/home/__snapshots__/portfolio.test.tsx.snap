// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portfolio /> <Porfolio.Image /> renders an image with populated data from props 1`] = `
<img
  alt="Gretsch Geeks icon"
  class="chakra-image css-1dzvt9d"
  src="/images/gretsch-logo.jpg"
/>
`;

exports[`<Portfolio /> <Portfolio.Button /> renders a button with text 1`] = `
<button
  class="chakra-button css-1efne3i"
  type="button"
>
  View
</button>
`;

exports[`<Portfolio /> <Portfolio.Item /> renders a button and calls openFeature() on click 1`] = `
<div
  class="css-79elbk"
  data-testid="portfolio-item"
  role="button"
>
  <div
    class="css-1xhj18k"
    role="group"
  >
    Children
  </div>
</div>
`;

exports[`<Portfolio /> <Portfolio.Text /> renders with populated data from props 1`] = `
<p
  class="chakra-text css-1lzpwia"
>
  An example E-commerce buy-sell platform built with modern React, Apollo 3 and Keystone.js as a headless CMS.
</p>
`;

exports[`<Portfolio /> <Portfolio.Title /> renders with populated data from props 1`] = `
<h3
  class="chakra-heading css-1l30lph"
>
  Gretsch Geeks
</h3>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders with a border-bottom if viewport width < 768px and projectsCount is 1 1`] = `
<div
  class="css-1yecze6"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders with a border-bottom if viewport width < 768px, projectsCount is 2 and it is not the last project 1`] = `
<div
  class="css-1yecze6"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders with a border-bottom if viewport width > 768px and projectsCount is < 3 1`] = `
<div
  class="css-1yecze6"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders with a border-bottom if viewport width > 768px, it is not in the first row, and it is the last one to be rendered before wrapping to a new row 1`] = `
<div
  class="css-1yecze6"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders without a border-bottom if viewport width < 768px, projectsCount is 2 and it is the last project 1`] = `
<div
  class="css-txocgk"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;

exports[`<Portfolio.Inner /> <Portfolio.Inner /> renders without a border-bottom if viewport width > 768px, projectsCount is 3 and it is on the bottom row 1`] = `
<div
  class="css-txocgk"
>
  <div
    class="css-2lzsxm"
  />
  <button
    class="chakra-button css-1efne3i"
    type="button"
  >
    View
  </button>
</div>
`;
